tfhe/src/keycache/mod.rs

```diff
@@ -155,63 +155,75 @@ pub mod utils {
         K: From<P> + Clone,
     {
         pub fn get(&self, param: P) -> SharedKey<K> {
-            self.with_key(param, |k| k.clone())
+            self.get_with_closure(param, &mut K::from)
         }
+    }
 
-        pub fn with_key<F, R>(&self, param: P, f: F) -> R
+    impl<P, K, S> KeyCache<P, K, S>
+    where
+        P: Copy + PartialEq + NamedParam,
+        S: PersistentStorage<P, K>,
+    {
+        pub fn get_with_closure<C: FnMut(P) -> K>(
+            &self,
+            param: P,
+            key_gen_closure: &mut C,
+        ) -> SharedKey<K> {
+            self.with_closure(param, key_gen_closure)
+        }
+
+        pub fn with_closure<C>(&self, param: P, key_gen_closure: &mut C) -> SharedKey<K>
         where
-            F: FnOnce(&SharedKey<K>) -> R,
+            C: FnMut(P) -> K,
         {
             let load_from_persistent_storage = || {
                 // we check if we can load the key from persistent storage
                 let persistent_storage = &self.persistent_storage;
                 let maybe_key = persistent_storage.load(param);
                 maybe_key.map_or_else(
                     || {
-                        let key = K::from(param);
+                        let key = key_gen_closure(param);
                         persistent_storage.store(param, &key);
                         key
                     },
                     |key| key,
                 )
             };
 
-            let try_load_from_memory_and_init = || {
-                // we only hold a read lock for a short duration to find the key
-                let maybe_shared_cell = {
-                    let memory_storage = self.memory_storage.read().unwrap();
-                    memory_storage
-                        .iter()
-                        .find(|(p, _)| *p == param)
-                        .map(|param_key| param_key.1.clone())
+            let try_load_from_memory_and_init = || -> Result<_, ()> {
+                let maybe_shared_key = {
+                    let mut res = None;
+                    let memory_iter = self.memory_storage.read().unwrap().clone().into_iter();
```

Comment:
clippy forced me to do it.

Reply:
just iter() vs clone().into_iter() does not compile ?

Reply:
Yes because of possible deadlock due the for loop.

Reply:
I don't understand how it can deadlock 😵‍💫 

the previous code is 

```rust
                let maybe_shared_cell = {
                    let memory_storage = self.memory_storage.read().unwrap();
                    memory_storage
                        .iter()
                        .find(|(p, _)| *p == param)
                        .map(|param_key| param_key.1.clone())
                };
```

i.e. we drop the lock right at the end of the block, the code above is rejected by clippy with what error message ?

Reply:
Here is the working implementation of what you've suggested previousely:
```rust
let maybe_shared_key = {
                    let mut res = None;
                    for (p, key) in self.memory_storage.read().unwrap().iter() {
                        if *p == param {
                            res = Some(key.clone());
                            break;
                        }
                    }
                    res
                };
```
And here's the associated error:
```rust
rror: temporary with significant `Drop` in `for` loop condition will live until the end of the `for` expression
   --> tfhe/src/keycache/mod.rs:197:37
    |
197 |                     for (p, key) in self.memory_storage.read().unwrap().iter() {
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
202 |                     }
    |                      - temporary lives until here
    |
    = note: this might lead to deadlocks or other unexpected behavior
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#significant_drop_in_scrutinee
    = note: `-D clippy::significant-drop-in-scrutinee` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::significant_drop_in_scrutinee)]`

```

Reply:
yes I meant the previous code ? why was clippy complaining ?

Reply:
this one https://github.com/zama-ai/tfhe-rs/pull/664#discussion_r1400690128

Reply:
Clippy wasn't complaining with the previous code, it's the refactor suggestion you made that triggered clippy.